<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Nonparametric Variable Importance for Differential Methylation Analysis • methyvim</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">methyvim</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="/index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nhejazi/methyvim">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    

    
    
<div class="contents">
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="rmethyvim" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rmethyvim" class="anchor"></a>R/<code>methyvim</code>
</h1></div>

<blockquote>
<p>Nonparametric variable importance analysis for the assessment of differential methylation</p>
</blockquote>
<p><strong>Author:</strong> <a href="http://nimahejazi.org">Nima Hejazi</a></p>
<hr>
<div id="description" class="section level2">
<h2 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h2>
<p><code>methyvim</code> is an R package that provides facilities for differential methylation analysis based on <em>variable importance measures</em> (VIMs), a class of statistical target parameters that arise in causal inference. The statistical techniques implemented rely on the use of targeted minimum loss-based estimation (TMLE) to assess two particular VIMs: (1) the “local” average treatment effect for discrete exposures, and (2) a nonparametric variable importance measure for continuous exposures. Within this framework, these methods allow differential methylation effects to be estimated in an assumption-free manner, on a pre-screened set of genomic sites measured by DNA methylation assays. As the statistical algorithm uses a multi-stage approach, multiple testing corrections are made using a modified marginal Benjamini &amp; Hochberg step-up False Discovery Rate controlling procedure for multi-stage analyses (FDR-MSA). In order to allow the user significant flexibility with respect to the scientific questions posed, the procedure estimates one of the appropriate VIMs for each of a reduced set of genomic sites, using screening procedures to identify a reduced set from the full data when a reduced set is not pre-specified. While making allowance for the user to specify the reduced set of genomic sites to be tested (e.g., those related to a particular gene of interest), this package comes equipped with screening procedures to facilitate the initial reduction of genomic sites, the two most noteworthy of these being a procedure based on the well-known R package <a href="https://bioconductor.org/packages/release/bioc/html/limma.html"><code>limma</code></a> and an extension of data-adaptive statistical target parameters for multiple testing (implemented in the R package <a href="https://github.com/wilsoncai1992/data.adapt.multi.test"><code>DA.Test</code></a>).</p>
<hr>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<!--
For standard use, install from [Bioconductor](https://bioconductor.org):

```r
source("https://bioconductor.org/biocLite.R")
biocLite("methyvim")
```
-->
<p>Install the most recent <em>stable release</em> from GitHub via <a href="https://www.rstudio.com/products/rpackages/devtools/"><code>devtools</code></a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"nhejazi/methyvim"</span>)</code></pre></div>
<p>To contribute, install the <em>development version</em> from GitHub via <a href="https://www.rstudio.com/products/rpackages/devtools/"><code>devtools</code></a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"nhejazi/methyvim"</span>, <span class="dt">ref =</span> <span class="st">"develop"</span>)</code></pre></div>
<hr>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is a basic example which shows you how to solve a common problem:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## basic example code</code></pre></div>
<hr>
</div>
<div id="issues" class="section level2">
<h2 class="hasAnchor">
<a href="#issues" class="anchor"></a>Issues</h2>
<p>If you encounter any bugs or have any specific feature requests, please <a href="https://github.com/nhejazi/methyvim/issues">file an issue</a>.</p>
<hr>
</div>
<div id="contributions" class="section level2">
<h2 class="hasAnchor">
<a href="#contributions" class="anchor"></a>Contributions</h2>
<p>It is our hope that <code>methyvim</code> will grow to be widely adopted as a tool for the nonparametric assessment of variable importance in studies of differential methylation. To that end, contributions are very welcome, though we ask that interested contributors consult our <a href="https://github.com/nhejazi/methyvim/blob/master/CONTRIBUTING.md"><code>contribution guidelines</code></a> prior to submitting a pull request.</p>
<hr>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>After using the <code>methyvim</code> R package, please cite it:</p>
<pre><code>    @article{hejazi2017methyvim,
      doi = {},
      url = {},
      year  = {2017},
      month = {},
      publisher={Faculty of 1000 Ltd},
      volume = {},
      author = {Hejazi, Nima S. and Hubbard, Alan E. and {van der Laan},
                Mark J.},
      title = {methyvim: Nonparametric Assessment of Variable Importance for
               Differential Methylation Analysis},
      journal = {F1000Research}
    }</code></pre>
<hr>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p><a href="http://www.targetedlearningbook.com">Mark J. van der Laan &amp; Sherri Rose. <em>Targeted Learning: Causal Inference for Observational and Experimental Data</em>, 2011.</a></p></li>
<li><p><a href="http://www.math-info.univ-paris5.fr/~chambaz/Papiers/chambazNeuvialvanderLaan_EJS2012.pdf">Antoine Chambaz, Pierre Neuvial, &amp; Mark J. van der Laan. “Estimation of a non-parametric variable importance measure of a continuous exposure”, <em>Electronic Journal of Statistics</em>, 2012.</a></p></li>
<li><p><a href="http://biostats.bepress.com/ucbbiostat/paper239/">Catherine Tuglus &amp; Mark J. van der Laan. “FDR controlling procedures for multi-stage analyses”, <em>UC Berkeley Division of Biostatistics Working Paper Series</em>, 2008.</a></p></li>
<li><p><a href="https://www.overleaf.com/5660573pjjrxh#/25678897/">Weixin Cai, Nima S. Hejazi, &amp; Alan E. Hubbard. “Data-adaptive statistics for multiple testing in high-dimensional settings”, <em>In preparation</em>, 2017.</a></p></li>
<li><p><a href="http://www.statsci.org/smyth/pubs/ebayes.pdf">Gordon K. Smyth. “Linear models and empirical Bayes methods for assessing differential expression in microarray experiments.” <em>Statistical Applications in Genetics and Molecular Biology</em>, 3(1), 2004.</a></p></li>
</ul>
<hr>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>© 2017 <a href="http://nimahejazi.org">Nima S. Hejazi</a>, <a href="http://sph.berkeley.edu/alan-hubbard">Alan E. Hubbard</a>, <a href="https://www.stat.berkeley.edu/~laan/">Mark J. van der Laan</a></p>
<p>The contents of this repository are distributed under the MIT license. See file <code>LICENSE</code> for details.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2 class="hasAnchor">
<a href="#sidebar" class="anchor"></a>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/nhejazi/methyvim">https://​github.com/​nhejazi/​methyvim</a>
</li>
<li>Report a bug at <br><a href="https://github.com/nhejazi/methyvim/issues">https://​github.com/​nhejazi/​methyvim/​issues</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE.html">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Nima Hejazi <br><small class="roles"> Author, maintainer </small> </li>
<li>Mark van der Laan <br><small class="roles"> Author </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/nhejazi/methyvim"><img src="https://travis-ci.org/nhejazi/methyvim.svg?branch=master" alt="Travis-CI Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/nhejazi/methyvim"><img src="https://ci.appveyor.com/api/projects/status/github/nhejazi/methyvim?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://codecov.io/github/nhejazi/methyvim?branch=master"><img src="https://img.shields.io/codecov/c/github/nhejazi/methyvim/master.svg" alt="Coverage Status"></a></li>
<li><a href="http://www.repostatus.org/#wip"><img src="http://www.repostatus.org/badges/latest/wip.svg" alt="Project Status: WIP - Initial development is in progress, but there has not yet been a stable, usable release suitable for the public."></a></li>
<li><a href="http://opensource.org/licenses/MIT"><img src="http://img.shields.io/badge/license-MIT-brightgreen.svg" alt="MIT license"></a></li>
</ul>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nima Hejazi, Mark van der Laan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
