\documentclass[9pt,a4paper,]{extarticle}

\usepackage{f1000_styles}

\usepackage[pdfborder={0 0 0}]{hyperref}

\usepackage[numbers]{natbib}
\bibliographystyle{unsrtnat}


%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}

% disable code chunks background
%\renewenvironment{Shaded}{}{}

% disable section numbers
\setcounter{secnumdepth}{0}

\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}



\usepackage{amsthm}
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}{Lemma}
\theoremstyle{definition}
\newtheorem{definition}{Definition}
\newtheorem{corollary}{Corollary}
\newtheorem{proposition}{Proposition}
\theoremstyle{definition}
\newtheorem{example}{Example}
\theoremstyle{definition}
\newtheorem{exercise}{Exercise}
\theoremstyle{remark}
\newtheorem*{remark}{Remark}
\newtheorem*{solution}{Solution}
\begin{document}
\pagestyle{front}

\title{\texttt{methyvim}: Targeted and Model-free Differential Methylation Analysis in R}

\author[1,2]{Nima S. Hejazi\thanks{\ttfamily nhejazi@berkeley.edu}}
\author[1]{Rachael V. Phillips}
\author[1]{Alan E. Hubbard}
\author[1,3]{Mark J. van der Laan}
\affil[1]{Group in Biostatistics, University of California, Berkeley}
\affil[2]{Center for Computational Biology, University of California, Berkeley}
\affil[3]{Department of Statistics, University of California, Berkeley}

\maketitle
\thispagestyle{front}

\begin{abstract}
We present \texttt{methyvim}, an R package implementing a general algorithm for the
nonparametric estimation of treatment effects on DNA methylation at CpG sites
throughout the genome, complete with straightforward statistical inference for
such estimates. The approach leverages variable importance measures derived
from statistical parameters arising causal inference, defined in such a manner
that they may be used to obtain targeted estimates of the relative importance
of individual CpG site with respect to a binary treatment assigned at the
phenotype level. The procedure implemented is computationally efficient,
incorporating a preliminary screening step to isolate a subset of sites for
which there is cursory evidence of differential methylation as well as a
unique multiple testing correction to control the False Discovery Rate with
the same rigor as if all sites were tested. This technique for analysis of
differentially methylated positions provides an avenue to incorporate flexible
state-of-the-art machine learning algorithms into the estimation of
differential methylation effects without the loss of interpretable statistical
inference.
\end{abstract}

\section*{Keywords}
DNA methylation, differential methylation, epigenetics, causal inference, statistical variable importance, targeted minimum loss-based estimation, machine learning


\clearpage
\pagestyle{main}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

DNA methylation is a fundamental epigenetic process known to play an important
role in the regulation of gene expression. DNA methylation occurs at CpG sites
and involves the addition of a methyl group (\(\text{CH}_3\)) to the fifth carbon
of the cytosine ring structure to form 5-methylcytosine. Numerous biological and
medical studies have implicated DNA methylation as playing a role in disease and
development \citep{robertson2005dna}. Perhaps unsurprisingly then, biotechnologies
have been developed to rigorously probe the molecular mechanisms of this
epigenetic process. Modern assays, like the Illumina \emph{Infinium} Methylation chip
technology, allow for quantitative interrogation of DNA methylation at a large
set of CpG sites scattered across the genome at single-nucleotide resolution;
moreover, much effort has been invested, by the bioinformatics community, in the
development of tools for properly removing technological effects that may
contaminate biological signatures measured by such assays
\citep[dedeurwaerder2013comprehensive]{fortin2014functional}. Despite these
advances in both biological and bioninformatical techniques, most statistical
methods available for differential analysis of data produced by such assays rely
on over-simplified models that fail to incorporate many of the recent advances
in machine learning. Generally, typical statistical estimation approaches do not
readily extend to such high-dimensional data without restrictive modeling
assumptions and the use of inferentially costly hypothesis testing corrections.
When these standard assumptions are violated, estimates of the population-level
effect of an exposure or treatment may suffer from large bias. Data-adaptive
estimation procedures that utilize machine learning can control for confounding
even in high-dimensional settings; however, interpretable statistical inference
(i.e., confidence intervals and hypothesis tests) resulting from these
data-adaptive estimates is challenging to obtain \citep{libbrecht2015machine}.

In this paper, we briefly present an alternative to such statistical analysis
approaches in the form of a nonparametric estimation procedure that provides
simple and readily interpretable statistical inference, discussing at length a
recent implementation of the methodology in the \texttt{methyvim} R package. Inspired
by recent advances in statistical causal inference and machine learning, we
provide a computationally efficient technique for obtaining targeted estimates
of nonparametric \emph{variable importance measures} (VIMs) \citep{vdl2006statistical},
estimated at a set of pre-screened CpG sites, controlling for the False
Discovery Rate (FDR) as if all sites were tested. Under standard assumptions
(e.g., identifiability, strong ignorability) \citep{pearl2009causality}, targeted
minimum loss-based estimators of regular asymptotically linear estimators have
sampling distributions that are asymptotically normal, allowing for reliable
point estimation and the construction of Wald-style confidence intervals
\citep[\citet{vdl2018targeted}]{vdl2011targeted}. As counterfactuals are contrary-to-fact --
that is, defined in terms of missing data -- in the context of DNA methylation,
we define the counterfactual outcomes under a binary treatment as the observed
methylation (whether Beta- or M-) value a CpG site would have if all subjects
were given the treatment and the methylation value a CpG site would have if
treatment were withheld from all subjects. Although these counterfactual
outcomes are, of course, impossible to observe, they have statistical analogs
that may be reliably estimated from observed data \citep{pearl2009causality}. We
describe an algorithm that incorporates the estimation of VIMs using \emph{targeted
minimum loss-based estimation} (\textbf{TMLE}) \citep{vdl2006targeted}, deferring detailed
description and analysis of the statistical methodology to work outside the
scope of the present manuscript. This methodology assesses the individual
importance of a given CpG site by utilizing state-of-the-art machine learning
algorithms in the estimation of a targeted VIM, initially proposed and explored
in \citep{tuglus2008targeted}. In the presnet work, we focus on a software package,
\texttt{methyvim}, that implements a variant of this methodology, specifically tailored
to differential methylation analysis.

For a general discussion of the framework of targeted minimum loss-based
estimation and the role this approach plays in statistical causal inference, the
interested reader is invited to consult \citet{vdl2011targeted} and \citet{vdl2018targeted}.
For a more general introduction to (statistical) causal inference,
\citet{pearl2009causality} and \citet{hernan2018causal} may be of interest.

\hypertarget{methods}{%
\section{Methods}\label{methods}}

\hypertarget{data-structure}{%
\subsection{Data Structure}\label{data-structure}}

We consider an observed data structure \(O = (W_k, A, Y_j)\), where \(Y_j, j = 1, \ldots J\) represents a given CpG site of interest, \(A \in \{0, 1\}\) represents a
binary phenotype-level treatment, and \(W_k, k = 1, \ldots K(j)\) is a matrix of
the observed values of CpG sites in the same neighborhood as \(j\). We consider
the case of observing \(n\) iid copies of \(O\) (i.e., \(O_1, \ldots, O_n\)), where we
take \(O \sim \mathcal{P} \in \mathcal{M}\), which is to say that \(O\) is governed
by an unknown probability distribution \(\mathcal{P}\), taken only to reside in a
nonparametric statistical model \(\mathcal{M}\) that places no restrictions on the
data-generating process. We consider having access to measurements on a large
number \(G\) of CpG sites (e.g., 850,000, as measured by the Illumina
MethylationEPIC BeadChip arrays), and as a matter of practicality, let \(J \leq G\), so that \(J\) indexes only those CpG sites that pass an arbitrary
pre-screening procedure; note that the equality is only obtained in the case
that all sites pass the pre-screening procedure. Further, we let \(K(j)\) index
the set of all CpG sites so that \(k\), the realization of \(K(j)\) for a given CpG
site \(j\), is a set of the CpG sites that neighbor the CpG site \(j\), where the
definition of a neighborhood is left as user-specified; thus, \(W_k\) is a matrix
of observed methylation values for the set of CpG sites \(k\) that neighbor a
particular target site \(j\). We consider estimating a target parameter (\(\psi_j\))
as a variable importance measure of the effect of changing a treatment \(A\) on
the methylation of a CpG site \(j\), accounting for the observed methylation at
the set of sites \(k\) that neighbor the site \(j\). After a complete run of the
procedure, we have access to \(j\) VIM estimates \(\psi_j\), each corresponding to
the degree to which a given CpG site \(j\) was responsive to a change in the
binary treatment \(A\), after controlling for any confounding exerted by its
neighboring CpG sites. When machine learning estimators are employed in
estimating \(\psi_j\), this produces a nonparametric variable importance measure
of differential methylation, allowing for the identification of differentially
methylated positions (DMPs).

\hypertarget{implementation}{%
\subsection{Implementation}\label{implementation}}

The core functionality of this package is made available via the eponymous
\texttt{methyvim} function, which implements a statistical algorithm designed to
compute targeted estimates of VIMs, defined in such a way that the VIMs
represent parameters of scientific interest in computational biology
experiments; moreover, these VIMs are defined such that they may be estimated in
a manner that is very nearly assumption-free, that is, within a \emph{fully
nonparametric statistical model}. \textbf{The statistical algorithm consists in
several major steps}:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Pre-screening of genomic sites is used to isolate a subset of sites for
  which there is cursory evidence of differential methylation. For the sake of
  computational feasibility, targeted minimum loss-based estimates of VIMs are
  computed only for this subset of sites. Currently, the available screening
  approach adapts core routines from the
  \href{http://bioconductor.org/packages/limma}{\texttt{limma}} R
  package. Future releases will support functionality from other packages
  (e.g., \href{https://CRAN.R-project.org/package=randomForest}{\texttt{randomForest}},
  \href{https://CRAN.R-project.org/package=tmle.npvi}{\texttt{tmle.npvi}}). Following the
  style of the function for performing screening via \texttt{limma}, users may write
  their own screening functions and are invited to contribute such functions to
  the core software package by opening pull requests at the GitHub repository.
\item
  Nonparametric estimates of VIMs, for the specified target parameter, are
  computed at each of the CpG sites passing the screening step. The VIMs are
  defined in such a way that the estimated effects is of an exposure/treatment
  on the methylation status of a target CpG site, controlling for the observed
  methylation status of the neighbors of that site. Currently, routines are
  adapted from the \href{https://CRAN.R-project.org/package=tmle}{\texttt{tmle}} R package.
  Future releases will support doubly-robust estimates of these VIMs (via the
  \href{https://cran.r-project.org/package=drtmle}{\texttt{drtmle}} package)
  and add parameters for continuous treatments/exposures (via the
  \href{https://CRAN.R-project.org/package=tmle.npvi}{\texttt{tmle.npvi}} and
  \href{https://github.com/nhejazi/txshift}{\texttt{txshift}} R packages).
\item
  Since pre-screening is performed prior to estimating VIMs, we make use of a
  multiple testing correction uniquely suited to such settings. Due to the
  multiple testing nature of the estimation problem, a variant of the Benjamini
  and Hochberg procedure for controlling the False Discovery Rate (FDR) is
  applied \citep{benjamini1995controlling}. Specifically, we apply a modified
  marginal Benjamini and Hochberg step-up False Discovery Rate controlling
  procedure for multi-stage analyses (FDR-MSA), which has established
  theoretical guarantees to control the FDR at the same rate as if all sites
  were tested (i.e., without screening) \citep{tuglus2009modified}.
\end{enumerate}

\hypertarget{parameters-of-interest}{%
\paragraph{Parameters of Interest}\label{parameters-of-interest}}

For the CpG sites that pass the pre-screening step, a user-specified target
parameter of interest is estimated independently at each site. \emph{In all cases,
an estimator of the parameter of interest is constructed via targeted minimum
loss-based estimation}.

For discrete-valued treatments or exposures:

\begin{itemize}
\item
  The average treatment effect (ATE): The effect of a binary exposure or
  treatment on the observed methylation at a target CpG site is estimated,
  controlling for the observed methylation at all other CpG sites in the same
  neighborhood as the target site, based on an additive form. Often denoted
  \(\psi_0 = \psi_0(1) - \psi_0(0)\), the parameter estimate represents the
  additive difference in methylation that would have been observed at the
  target site had all observations received the treatment versus the
  counterfactual under which none received the treatment.
\item
  The relative risk (RR): The effect of a binary exposure or treatment on the
  observed methylation at a target CpG site is estimated, controlling for the
  observed methylation at all other CpG sites in the same neighborhood as the
  target site, based on a geometric form. Often denoted,
  \(\psi_0 = \frac{\psi_0(1)}{\psi_0(0)}\), the parameter estimate represents the
  multiplicative difference in methylation that would have
  been observed at the target site had all observations received the treatment
  versus the counterfactual under which none received the treatment.
\end{itemize}

Estimating the VIM corresponding to the parameters above, for discrete-valued
treatments or exposures, requires two separate regression steps: one for the
treatment mechanism (propensity score) and one for the outcome regression.
Technical details on the nature of these regressions are discussed in
\citet{hernan2018causal}, and details for estimating these regressions in the
framework of targeted minimum loss-based estimation are discussed in
\citet{vdl2011targeted}.

Support for continuous-valued treatments or exposures is \emph{planned but not yet
available}. Future releases will allow users to assess continuous-valued
treatments by relying on parameters estimable through implementations available
in the following software packages:

\begin{itemize}
\item
  \texttt{tmle.npvi} - A nonparametric variable importance measure (NPVI)
  \citep{chambaz2012estimation}. The effect of a continuous-valued exposure or
  treatment (the observed methylation at a target CpG site) on an outcome of
  interest is estimated, controlling for the observed methylation value at all
  other CpG sites in the same neighborhood as the target site. This uses a
  parameter that compares values of the treatment against a user-specified
  reference value taken to be the null. In particular, the implementation to be
  provided is designed to assess the effect of differential methylation at the
  target CpG site on an outcome of interest (e.g., survival), providing a
  nonparametric evaluation of the impact of methylation at the target site.
\item
  \texttt{txshift} - Variable importance using a causal parameter defined as the
  counterfactual outcome under a posited shift of a continuous-value treatment
  of interest using stochastic intervention policies \citep[
  \citet{diaz2018stochastic}, \citet{hejazi2018txshift}]{munoz2012population}. The value of an outcome of interest
  under an unobserved value of the (continuous-valued) treatment, specified
  through a user-provided \emph{additive shift} of the observed treatment, may be
  data-adaptively estimated. This allows for the effect of changes/shifts in the
  observed methylation at a CpG site on an outcome of interest to be evalauted,
  providing a nonparametric evalution of the relative importance of a particular
  target CpG site with respect to another variable measured in the same study.
\end{itemize}

\hypertarget{class-methytmle}{%
\paragraph{\texorpdfstring{Class \texttt{methytmle}}{Class methytmle}}\label{class-methytmle}}

We have adopted a class \texttt{methytmle} to help organize the functionality within
this package. The \texttt{methytmle} class builds upon the \texttt{GenomicRatioSet} class
provided by the \texttt{minfi} package so all of the slots of \texttt{GenomicRatioSet} are
contained in a \texttt{methytmle} object. The new class introduced in the \texttt{methyvim}
package includes several new slots:

\begin{itemize}
\item
  \texttt{call} - the form of the original call to the \texttt{methyvim} function.
\item
  \texttt{screen\_ind} - indices identifying CpG sites that pass the screening process.
\item
  \texttt{clusters} - non-unique IDs corresponding to the manner in wich sites are
  treated as neighbors. These are assigned by genomic distance (bp) and respect
  chromosome boundaries (produced via a call to \texttt{bumphunter::clusterMaker}).
\item
  \texttt{var\_int} - the treatment/exposure status for each subject. Currently, these
  must be binary, due to the definition of the supported targeted parameters.
\item
  \texttt{param} - the name of the target parameter from which the estimated VIMs are
  defined.
\item
  \texttt{vim} - a table of statistical results obtained from estimating VIMs for
  each of the CpG sites that pass the screening procedure.
\item
  \texttt{ic} - the measured array values for each of the CpG sites passing the
  screening, transformed into influence curve space based on the chosen target
  parameter.
\end{itemize}

We refer the reader to the package vignette, ``\texttt{methyvim}: Targeted Data-Adaptive
Estimation and Inference for Differential Methylation Analysis,'' included in any
distribution of the software package, for further details.

\hypertarget{operation}{%
\subsection{Operation}\label{operation}}

A standard computer with the latest version of R and Bioconductor 3.6 installed
will handle applications of the \texttt{methyvim} package.

\hypertarget{use-cases}{%
\section{Use Cases}\label{use-cases}}

To examine the practical applications and the full set of utilities of the
\texttt{methyvim} package, we will use a publicly available example data set produced
by the Illumina 450K array, from the \texttt{minfiData} R package.

\hypertarget{preliminaries-setting-up-the-data}{%
\paragraph{Preliminaries: Setting up the Data}\label{preliminaries-setting-up-the-data}}

We begin by loading the package and the data set. After loading the data, which
comes in the form of a raw \texttt{MethylSet} object, we perform some further
processing by mapping to the genome (with \texttt{mapToGenome}) and converting the
values from the methylated and unmethylated channels to Beta-values
(via \texttt{ratioConvert}). These two steps together produce an object of class
\texttt{GenomicRatioSet}, provided by the \texttt{minfi} package.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{library}\NormalTok{(minfiData))}
\KeywordTok{data}\NormalTok{(MsetEx)}
\NormalTok{mset <-}\StringTok{ }\KeywordTok{mapToGenome}\NormalTok{(MsetEx)}
\NormalTok{grs <-}\StringTok{ }\KeywordTok{ratioConvert}\NormalTok{(mset)}
\NormalTok{grs}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: GenomicRatioSet 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
\end{verbatim}

We can create an object of class \texttt{methytmle} from any \texttt{GenomicRatioSet} object
simply invoking the S4 class constructor \texttt{.methytmle}:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(methyvim)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## methyvim v1.3.1: Targeted Variable Importance for Differential Methylation Analysis
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{grs_mtmle <-}\StringTok{ }\KeywordTok{.methytmle}\NormalTok{(grs)}
\NormalTok{grs_mtmle}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: methytmle 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
## Target Parameter: 
## Results: 
## Object of class "data.frame"
## data frame with 0 columns and 0 rows
\end{verbatim}

Additionally, a \texttt{GenomicRatioSet} can be created from a matrix with the
function \texttt{makeGenomicRatioSetFromMatrix} provided by the \texttt{minfi} package.

\hypertarget{differential-methylation-analysis}{%
\paragraph{Differential Methylation Analysis}\label{differential-methylation-analysis}}

For this example analysis, we'll treat the condition of the patients as the
exposure/treatment variable of interest. The \texttt{methyvim} function requires that
this variable either be \texttt{numeric} or easily coercible to \texttt{numeric}. To
facilitate this, we'll simply convert the covariate (currently a \texttt{character}):

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{var_int <-}\StringTok{ }\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(}\KeywordTok{as.factor}\NormalTok{(}\KeywordTok{colData}\NormalTok{(grs)}\OperatorTok{$}\NormalTok{status)) }\OperatorTok{-}\StringTok{ }\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbf{n.b.}, the re-coding process results in ``normal'' patients being assigned a
value of 1 and cancer patients a 0.

Now, we are ready to analyze the effects of cancer status on DNA methylation
using this data set. We proceed as follows with a targeted minimum loss-based
estimate of the Average Treatment Effect.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{methyvim_cancer_ate <-}\StringTok{ }\KeywordTok{methyvim}\NormalTok{(}\DataTypeTok{data_grs =}\NormalTok{ grs, }\DataTypeTok{var_int =}\NormalTok{ var_int,}
                                \DataTypeTok{vim =} \StringTok{"ate"}\NormalTok{, }\DataTypeTok{type =} \StringTok{"Beta"}\NormalTok{, }\DataTypeTok{filter =} \StringTok{"limma"}\NormalTok{,}
                                \DataTypeTok{filter_cutoff =} \FloatTok{0.20}\NormalTok{, }\DataTypeTok{obs_per_covar =} \DecValTok{2}\NormalTok{,}
                                \DataTypeTok{parallel =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{sites_comp =} \DecValTok{250}\NormalTok{,}
                                \DataTypeTok{tmle_type =} \StringTok{"glm"}
\NormalTok{                               )}
\end{Highlighting}
\end{Shaded}

Note that we set the \texttt{obs\_per\_covar} argument to a relatively low value (just 2,
even though the recommended value, and default, is 20) for the purposes of this
example as the sample size is only 10. We do this only to exemplify the
estimation procedure and it is important to point out that such low values for
\texttt{obs\_per\_covar} will compromise the quality of inference obtained because this
setting directly affects the definition of the target parameter.

Further, note that here we apply the \texttt{glm} flavor of the \texttt{tmle\_type} argument,
which produces faster results by fitting models for the propensity score and
outcome regressions using a limited number of parametric models. By contrast,
the \texttt{sl} (for ``Super Learning'') flavor fits these two regressions using highly
nonparametric and data-adaptive procedures (i.e., via machine learning).
Obtaining the estimates via GLMs results in each of the regression steps
being less robust than if nonparametric regressions were used.

We can view a table of results by examining the \texttt{vim} slot of the produced
object, most easily displayed by simply printing the resultant object:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{methyvim_cancer_ate}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: methytmle 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
## Target Parameter: Average Treatment Effect
## Results: 
##                    lwr_ci       est_ate        upr_ci        var_ate
## cg14008030 -0.11956597277 -0.0314159619  0.0567340489 0.002022705230
## cg20253340 -0.08850636712 -0.0588661418 -0.0292259165 0.000228691940
## cg21870274 -0.09499057246 -0.0291189817  0.0367526091 0.001129494604
## cg17308840 -0.04626018088 -0.0071524518  0.0319552773 0.000398119136
## cg00645010 -0.02677328296 -0.0134655543 -0.0001578256 0.000046099449
## cg27534567  0.06745648461  0.1157118536  0.1639672225 0.000606148645
## cg08258224  0.13650451332  0.3050884951  0.4736724770 0.007398104677
## cg20275697 -0.30210263597 -0.1231299608  0.0558427144 0.008337988977
## cg24373735 -0.04283533168  0.0076666975  0.0581687266 0.000663904349
## cg12445832 -0.06082410546  0.0150395574  0.0909032203 0.001498150599
## cg01097950 -0.04392159158  0.0658323796  0.1755863507 0.003135655504
## cg01782097 -0.01082071508  0.0010232901  0.0128672954 0.000036516155
##                                                      pval n_neighbors
## cg14008030 0.48484680819856229572195616128738038241863251           0
## cg20253340 0.00009917426050587487018107835101687896894873           0
## cg21870274 0.38625371823607712595816110479063354432582855           2
## cg17308840 0.71999433317147576438799205789109691977500916           2
## cg00645010 0.04734006904146480232409288646522327326238155           2
## cg27534567 0.00000260293555116907031197386393484016764432           1
## cg08258224 0.00038959136906232085580809032521187873499002           1
## cg20275697 0.17751545277429223168574878854997223243117332           0
## cg24373735 0.76604893436462262457098404411226511001586914           0
## cg12445832 0.69760217179702999068524604808771982789039612           0
## cg01097950 0.23973765213124559325663653908122796565294266           0
## cg01782097 0.86553022244577304533663664187770336866378784           1
##            n_neighbors_control max_cor_neighbors
## cg14008030                   0                NA
## cg20253340                   0                NA
## cg21870274                   1        0.94435796
## cg17308840                   1        0.94435796
## cg00645010                   2        0.52368097
## cg27534567                   0        0.93629683
## cg08258224                   0        0.93629683
## cg20275697                   0                NA
## cg24373735                   0                NA
## cg12445832                   0                NA
## cg01097950                   0                NA
## cg01782097                   1       -0.39410834
##  [ reached getOption("max.print") -- omitted 238 rows ]
\end{verbatim}

Finally, we may compute FDR-corrected p-values, by applying a modified procedure
for controlling the False Discovery Rate for multi-stage analyses (FDR-MSA)
\citep{tuglus2009modified}. We do this by simply applying the \texttt{fdr\_msa} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fdr_p <-}\StringTok{ }\KeywordTok{fdr_msa}\NormalTok{(}\DataTypeTok{pvals =}\NormalTok{ methyvim_cancer_ate}\OperatorTok{@}\NormalTok{vim}\OperatorTok{$}\NormalTok{pval,}
                 \DataTypeTok{total_obs =} \KeywordTok{nrow}\NormalTok{(methyvim_cancer_ate))}
\end{Highlighting}
\end{Shaded}

Having explored the results of our analysis numerically, we now proceed to use
the visualization tools provided with the \texttt{methyvim} R package to further
enhance our understanding of the results.

\hypertarget{visualization-of-results}{%
\paragraph{Visualization of Results}\label{visualization-of-results}}

While making allowance for users to explore the full set of results produced by
the estimation procedure (by way of exposing these directly to the user), the
\texttt{methyvim} package also provides \emph{three} (3) visualization utilities that
produce plots commonly used in examining the results of differential methylation
analyses.

A simple call to \texttt{plot} produces side-by-side histograms of the raw p-values
computed as part of the estimation process and the corrected p-values obtained
from using the FDR-MSA procedure.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(methyvim_cancer_ate, }\DataTypeTok{type =} \StringTok{"raw_pvals"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-pvals-raw-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(methyvim_cancer_ate, }\DataTypeTok{type =} \StringTok{"fdr_pvals"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-pvals-fdr-1} \end{center}

\textbf{Remark:} The plots displayed above may also be generated as side-by-side
histograms in a single plot object. This is the default for the \texttt{plot} method
and may easily be invoked by specifying no additional arguments to the \texttt{plot}
function, unlike in the above.

While histograms of the p-values may be generally useful in inspecting the
results of the estimation procedure, a more common plot used in examining the
results of differential methylation procedures is the volcano plot, which plots
the parameter estimate along the x-axis and \(-\text{log}_{10}(\text{p-value})\)
along the y-axis. We implement such a plot in the \texttt{methyvolc} function:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{methyvolc}\NormalTok{(methyvim_cancer_ate)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-volcano-1} \end{center}

The purpose of such a plot is to ensure that very low (possibly statistically
significant) p-values do not arise from cases of low variance. This appears to
be the case in the plot above (notice that most parameter estimates are near
zero, even in cases where the raw p-values are quite low).

Yet another popular plot for visualizing effects in such settings is the
heatmap, which plots estimates of the raw methylation effects (as measured by
the assay) across subjects using a heat gradient. We implement this in the
\texttt{methyheat} function:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{methyheat}\NormalTok{(methyvim_cancer_ate)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-heatmap-1} \end{center}

Invoking \texttt{methyheat} in this manner produces a plot of the top sites (\(25\), by
default) based on the raw p-value, using the raw methylation measures in the
plot. This uses the exceptional \texttt{superheat} R package \citep{barter2017superheat}.

\hypertarget{summary}{%
\section{Summary}\label{summary}}

Here we introduce the R package \texttt{methyvim}, an implementation of a general
algorithm for differential methylation analysis that allows for recent advances
in causal inference and machine learning to be leveraged in computational
biology settings. The estimation procedure produces straightforward statistical
inference and takes great care to ensure computationally efficiency of the
technique for obtaining targeted estimates of nonparametric variable importance
measures. The software package includes techniques for pre-screening a set of
CpG sites, controlling for the False Discovery Rate as if all sites were tested,
and for visualzing the results of the analyses in a variety of ways. The anatomy
of the software package is dissected and the design described in detail. The
\texttt{methyvim} R package is available via the Bioconductor project.

\hypertarget{software-availability}{%
\section{Software availability}\label{software-availability}}

Latest source code (development version): \url{https://github.com/nhejazi/methyvim}

Bioconductor (stable release): \url{https://bioconductor.org/packages/methyvim}

Archived source code as at time of publication:
\url{https://github.com/nhejazi/methyvim/releases/tag/f1000}

Documentation (development version): \url{https://code.nimahejazi.org/methyvim}

Software license: The MIT License, copyright Nima S. Hejazi

\hypertarget{author-contributions}{%
\section{Author contributions}\label{author-contributions}}

NH designed and implemented the software package, applied the tool to the use
cases presented, and co-drafted the present manuscript. RP helped in designing
the software and co-drafted the present manuscript. AH and ML served as advisors
for the development of this software and the general statistical algorithm it
implements.

\hypertarget{competing-interests}{%
\section{Competing interests}\label{competing-interests}}

No competing interests were disclosed.

\hypertarget{grant-information}{%
\section{Grant information}\label{grant-information}}

NH was supported in part by the National Library of Medicine of the National
Institutes of Health under Award Number T32-LM012417, by P42-ES004705, and
by R01-ES021369. RP was supported by P42-ES004705. The content of this work is
solely the responsibility of the authors and does not necessarily represent the
official views of the various funding sources and agencies.

\renewcommand\refname{References}
{\small\bibliography{references.bib}}

\end{document}
