\documentclass[9pt,a4paper,]{extarticle}

\usepackage{f1000_styles}

\usepackage[pdfborder={0 0 0}]{hyperref}

\usepackage[numbers]{natbib}
\bibliographystyle{unsrtnat}


%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}

% disable code chunks background
%\renewenvironment{Shaded}{}{}

% disable section numbers
\setcounter{secnumdepth}{0}

\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}



\usepackage{amsthm}
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}{Lemma}
\theoremstyle{definition}
\newtheorem{definition}{Definition}
\newtheorem{corollary}{Corollary}
\newtheorem{proposition}{Proposition}
\theoremstyle{definition}
\newtheorem{example}{Example}
\theoremstyle{definition}
\newtheorem{exercise}{Exercise}
\theoremstyle{remark}
\newtheorem*{remark}{Remark}
\newtheorem*{solution}{Solution}
\begin{document}
\pagestyle{front}

\title{\texttt{methyvim}: Targeted, Robust, and Model-free Analysis of Differential Methylation in R}

\author[1,2]{Nima S. Hejazi\thanks{\ttfamily nhejazi@berkeley.edu}}
\author[1]{Rachael V. Phillips}
\author[1]{Alan E. Hubbard}
\author[1,3]{Mark J. van der Laan}
\affil[1]{Group in Biostatistics, University of California, Berkeley}
\affil[2]{Center for Computational Biology, University of California, Berkeley}
\affil[3]{Department of Statistics, University of California, Berkeley}

\maketitle
\thispagestyle{front}

\begin{abstract}
We present \texttt{methyvim}, an R package implementing a general algorithm for the
nonparametric estimation of treatment effects on DNA methylation at CpG sites
throughout the genome, complete with straightforward statistical inference for
such estimates. The approach leverages variable importance measures derived
from statistical parameters arising in causal inference, defined in such a
manner that they may be used to obtain targeted estimates of the relative
importance of individual CpG sites with respect to a binary treatment assigned
at the phenotype level. The procedure implemented is computationally
efficient, incorporating a preliminary screening step to isolate a subset of
sites for which there is cursory evidence of differential methylation as well
as a unique multiple testing correction to control the False Discovery Rate
with the same rigor as if all sites were tested. This technique for analysis
of differentially methylated positions provides an avenue to incorporate
flexible state-of-the-art machine learning algorithms into the estimation of
differential methylation effects without the loss of interpretable statistical
inference.
\end{abstract}

\section*{Keywords}
DNA methylation, differential methylation, epigenetics, causal inference, statistical variable importance, machine learning, targeted minimum loss-based estimation


\clearpage
\pagestyle{main}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

DNA methylation is a fundamental epigenetic process known to play an important
role in the regulation of gene expression. DNA methylation mostly commonly
occurs at CpG sites and involves the addition of a methyl group (\(\text{CH}_3\))
to the fifth carbon of the cytosine ring structure to form 5-methylcytosine.
Numerous biological and medical studies have implicated DNA methylation as
playing a role in disease and development \citep{robertson2005dna}. Perhaps
unsurprisingly then, biotechnologies have been developed to rigorously probe the
molecular mechanisms of this epigenetic process. Modern assays, like the
Illumina \emph{Infinium} HumanMethylation BeadChip assay, allow for quantitative
interrogation of DNA methylation, at single-nucleotide resolution, across a
comprehensive set of CpG sites scattered across the genome; moreover, the
computational biology community has invested significant effort in the
development of tools for properly removing technological effects that may
contaminate biological signatures measured by such assays
\citep[\citet{dedeurwaerder2013comprehensive}]{fortin2014functional}. Despite these
advances in both biological and bioninformatical techniques, most statistical
methods available for differential analysis of data produced by such assays rely
on over-simplified models that do not readily extend to such high-dimensional
data structures without restrictive modeling assumptions and the use of
inferentially costly hypothesis testing corrections. When these standard
assumptions are violated, estimates of the population-level effect of an
exposure or treatment may suffer from large bias. What's more, reliance on
restrictive and misspecified statistical models naturally leads to biased effect
estimates that are not only misleading in assessing effect sizes but also result
in false discoveries as these biased estimates are subject to testing and
inferential procedures. Such predictably unreliable methods serve only to
produce findings that are later invalidated by replication studies and add still
further complexity to discovering biological targets for potential therapeutics.
Data-adaptive estimation procedures that utilize machine learning provide a way
to overcome many of the problems common in classical methods, controlling for
potential confounding even in high-dimensional settings; however, interpretable
statistical inference (i.e., confidence intervals and hypothesis tests) from
such data-adaptive estimates is challenging to obtain \citep{libbrecht2015machine}.

In this paper, we briefly present an alternative to such statistical analysis
approaches in the form of a nonparametric estimation procedure that provides
simple and readily interpretable statistical inference, discussing at length a
recent implementation of the methodology in the \texttt{methyvim} R package. Inspired
by recent advances in statistical causal inference and machine learning, we
provide a computationally efficient technique for obtaining targeted estimates
of nonparametric \emph{variable importance measures} (VIMs) \citep{vdl2006statistical},
estimated at a set of pre-screened CpG sites, controlling for the False
Discovery Rate (FDR) as if all sites were tested. Under standard assumptions
(e.g., identifiability, strong ignorability) \citep{pearl2009causality}, targeted
minimum loss-based estimators of regular asymptotically linear estimators have
sampling distributions that are asymptotically normal, allowing for reliable
point estimation and the construction of Wald-style confidence intervals
\citep[\citet{vdl2018targeted}]{vdl2011targeted}. In the context of DNA methylation studies,
we define the counterfactual outcomes under a binary treatment as the observed
methylation (whether Beta- or M-) values a CpG site would have if all subjects
were administered the treatment and the methylation values a CpG site would have
if treatment were withheld from all subjects. Although these counterfactual
outcomes are, of course, impossible to observe, they do have statistical analogs
that may be reliably estimated (i.e., identified) from observed data under a
small number of untestable assumptions \citep{pearl2009causality}. We describe an
algorithm that incorporates, in its final step, the use \emph{targeted minimum
loss-based estimators} (\textbf{TMLE}) \citep{vdl2006targeted} of a given VIM of interest,
though we defer rigorous and detailed descriptions of this aspect of the
statistical methodology to work outside the scope of the present manuscript
\citep[\citet{vdl2011targeted}, \citet{vdl2018targeted}]{vdl2006targeted}. The proposed methodology
assesses the individual importance of a given CpG site, as a proposed measure of
differential methylation, by utilizing state-of-the-art machine learning
algorithms in deriving targeted estimates and robust inference of a VIM, as
considered more broadly for biomarkers in \citet{bembom2009biomarker} and
\citet{tuglus2011targeted}. In the present work, we focus on the \texttt{methyvim} software
package, available through the Bioconductor project \citep[
\citet{huber2015orchestrating}]{gentleman2004bioconductor} for the R language and environment for statistical
computing \citep{R}, which implements a particular realization of this methodology
specifically tailored for the analysis and identification of differentially
methylated positions (DMPs).

For an extended discussion of the general framework of targeted minimum
loss-based estimation and detailed accounts of how this approach may be brought
to bear in developing answers to complex scientific problems through tatistical
and causal inference, the interested reader is invited to consult
\citet{vdl2011targeted} and \citet{vdl2018targeted}. For a more general introduction to causal
inference, \citet{pearl2009causality} and \citet{hernan2018causal} may be of interest.

\hypertarget{methods}{%
\section{Methods}\label{methods}}

\hypertarget{data-structure}{%
\subsection{Data Structure}\label{data-structure}}

We consider an observed data structure, on a single experimental subject (e.g.,
a patient), \(O = (W, A, (Y(j) : j))\), where \((Y(j) : j = 1, \ldots J)\) is the
set of CpG sites measured by the assay in question, \(A \in \{0, 1\}\) represents
a binary phenotype-level treatment, and \(W\) is a vector of the phenotype-level
baseline covariates that are potential confounders (e.g., age, sex). We consider
having access to measurements on a large number \(J\) of CpG sites (e.g.,
\(850,000\), as measured by the Illumina \emph{MethylationEPIC} BeadChip arrays).
Further, let \(K: j \to S_j\) be a procedure that assigns to a given CpG site \(j\)
a set of neighbors \(S_j\) -- that is, \(S_j\) is a collection of indices of the
neighbors of \(j\) just as \(j\) indexes the full set of CpG sites; moreover, since
\(Y(j)\) is the measured methylation value for a given CpG site \(j\), \(Y(S_j)\) is
the measured methylation values at the neighbors \(S_j\) of \(j\). Note that the
definition of a neighborhood \(\{j, S_j\}\) is left vague so as to facilitate the
use of user-specified strategies in implementation. We consider the case of
observing \(n\) iid copies of \(O\) (i.e., \(O_1, \ldots, O_n\)), where \(O \sim P \in \mathcal{M}\), which is to say that the random variable \(O\) is governed by an
unknown probability distribution \(P\), assumed only to reside in a nonparametric
statistical model \(\mathcal{M}\) that places no restrictions on the
data-generating process.

\hypertarget{variable-importance-measure}{%
\subsection{Variable Importance Measure}\label{variable-importance-measure}}

With the data structure above in hand, we let the estimand of interest be a
\(j\)-specific variable importance measure (VIM) \(\Psi_j(P)\), which is defined
through the \emph{true} (and unknown) probability distribution \(P\). As a motivating
example, consider the case where \(\Psi_j(P)\) is
\begin{equation}\label{vim_param}
  \Psi_j(P) = \mathbb{E}_P(\mathbb{E}_P(Y(j) \mid A = 1, W, Y(S_j)) -
    \mathbb{E}_P \mathbb{E}_P(Y(j) \mid A = 0, W, Y(S_j)),
\end{equation}
where \(Y(S_j)\) is the subvector of \(Y\) that contains the measured methylation
values of the \emph{neighboring sites} of \(j\) (but not site \(j\) itself). This
parameter of interest is a variant of the \emph{average treatment effect}, which has
been the subject of much attention in statistical causal inference \textbf{CITATION
NEEDED}. As a measure of variable importance, we seek to estimate this target
parameter (\(\Psi_j(P)\)), which quantifies the effect of changing a treatment \(A\)
on the methylation \(Y(j)\) of a CpG site \(j\), accounting for any potential
confounding from phenotype-level covariates \(W\) and observed methylation
\(Y(S_j)\) at the neighboring sites \(S_j\) of \(j\). We propose a procedure that
estimates this target parameter \(\Psi_j(P)\) across all CpG sites of interest
\(j: 1, \ldots, J\). When data-adaptive regression procedures (i.e., machine
learning) are employed in estimating \(\Psi_j(P)\), this produces a nonparametric
variable importance measure of differential methylation, allowing for the
identification of differentially methylated positions (DMPs) while avoiding many
assumptions common in the use of standard parametric regression procedures. We
propose estimating \(\Psi_j(P)\) via targeted minimum loss-based estimation, which
allows for data-adaptive regression procedures to be employed in a
straightforward manner.

\hypertarget{pre-screening-procedure}{%
\subsection{Pre-Screening Procedure}\label{pre-screening-procedure}}

As a matter of practicality, we consider only estimating the target VIM
\(\Psi_j(P)\) at a subset of CpG sites, so that \(j: 1, \ldots J\) does not, in fact
span the full set of assayed CpG sites. In order to determine this subset of CpG
sites, we propose the use of a pre-screening procedure, which need be nothing
more than a method for differential methylation analysis that is computationally
less demanding than the method proposed here. Formally, let the pre-screening
procedure \(\theta(j): j \to \{0, 1\}\), which is to say that \(\theta\) takes as
input a CpG site \(j\) and returns a binary decision rule of whether to include
CpG site \(j\) in a subset to be evaluated further or not. Details on the
implementation of pre-screening approaches is available in the sequel.

\hypertarget{reducing-neighbors-via-clustering}{%
\subsection{Reducing Neighbors via Clustering}\label{reducing-neighbors-via-clustering}}

Due to the data-adaptive nature of the regression procedures employed in
evaluating the target VIM \(\Psi_j(P)\) via targeted minimum loss-based
estimation, it is possible the a given CpG site \(j\) may have \emph{too many}
neighbors \(S_j\) to be controlled for in the estimation procedure. Heuristically,
the inclusion of too many neighbors when controlling for potential confounders
may lead to instability in the estimates produced. In such cases, we propose and
implement the use of a clustering technique (e.g., Partitioning Around Medoids)
to select a \emph{representative} subset of neighbors. Formally, there is likely no
best choice of a specific clustering algorithm, so we leave this aspect of the
proposed algorithm as flexible for the user \citep{kleinberg2003impossibility}.

\hypertarget{targeted-minimum-loss-based-estimation-and-inference}{%
\subsection{Targeted Minimum Loss-Based Estimation and Inference}\label{targeted-minimum-loss-based-estimation-and-inference}}

Given the choice of target parameter \(\Psi_j(P)\), we propose the use of targeted
minimum loss-based estimation (TMLE) to construct and evaluate an estimator
\(\psi_{n, j}\) of \(\Psi_j(P)\). In the case of our motivating example, where the
target VIM is the average treatment effect, we make use of a TML estimator of
this parameter, which has been implemented for a general case in the
\href{https://CRAN.R-project.org/package=tmle}{\texttt{tmle}} R package \citep{gruber2011tmle}.
Generally speaking, a TML estimator is constructed from a few simple components:
(1) an estimator of the propensity score \citep{rosenbaum1983central} (often denoted
by \(g(A, W)\)), (2) an estimator of the outcome regression (often denoted \(Q(A, W)\)), and (3) a targeting step applied to the initial estimator constructed from
the aforementioned components \citep{gruber2009targeted}. In the sequel, we describe
a few of the key mathematical properties of TML estimators relevant for
statistical inference, deferring extended discussion of the technical details to
\citet{vdl2011targeted} and \citet{vdl2018targeted}; moreover, as a matter of practicality, we
make use of the \href{https://CRAN.R-project.org/package=tmle}{\texttt{tmle}} R package of
\citet{gruber2011tmle} as a backend for computing estimators of relevant target
parameters in our software package, so the interested user is encouraged to
discuss the documentation of that software package for further details.

As noted previously, data-adaptive regression techniques may be used in the
construction of TML estimators. This is done by generating a stacked regression
model from a given set of machine learning algorithms via the Super Learner
algorithm \citep[breiman1996stacked, vdl2007super]{wolpert1992stacked}, which ensures
that the resulting ensemble learner satisifies important optimality properties
with respect to cross-validation \citep[\citet{dudoit2005asymptotics},
\citet{vdl2003unified}]{vdl2004asymptotic}. Importantly, TML estimators are well suited for inference,
having an asymptotically normal limiting distribution
\citep[\citet{vdl2006targeted}]{tsiatis2007semiparametric} \(\sqrt{n}(\psi_{n, j} - \Psi_j(P)) \sim N(0, \sigma_{\text{EIF}}^2)\), where \(\sigma_{\text{EIF}}^2\) is
the variance of the efficient influence function of the target parameter
evaluated at the observed data. This allows for confidence intervals to be
constructed and and hypothesis tests (i.e., \(H_{0, j}: \Psi_j(P) = 0\)) to be
performed in a straightforward manner.

\hypertarget{correction-for-multiple-testing}{%
\subsection{Correction for Multiple Testing}\label{correction-for-multiple-testing}}

Given that we seek to estimate \(\Psi_j(P)\) for a possibly large number of CpG
sites \((j: 1, \ldots J)\), the need to perform corrections for multiple testing
is clear. In order to curb the potential for false discoveries, we recommend the
use of the Benjamini and Hochberg procedure for controlling the False Discovery
Rate (FDR) \citep{benjamini1995controlling}; however, as the proposed procedure
involves a pre-screening step, naive application of the Benjamini and Hochberg
procedure (BH) is invalid -- instead, we rely on a modification of the procedure
to control the FDR, with established theoretical guarantees when pre-screening
is employed \citep{tuglus2009modified}. In brief, the modified marginal Benjamini and
Hochberg procedure to control the FDR under pre-screening works by applying the
standard BH procedure to a padded vector of p-values -- that is, letting \(J\) be
the number of CpG sites tested and \(K\) the number of CpG sites filtered out (so
that \(J + K = P\), where \(P\) is the original dimension of the genomic assay), the
modified marginal BH procedure is the application of the original BH procedure
to a vector of p-values, composed of the \(J\) p-values from performing \(J\)
hypothesis tests (of the form \(H_{0, j}: \Psi_j(P) = 0\)) and \(K\) additional
p-values automatically set to \(1\) (for the hypothesis tests not performed on
account of pre-screening). This procedure is guaranteed to control the FDR at
the same desired rate when pre-screening is performed whereas naive application
of the BH procedure fails to do so.

\hypertarget{implementation}{%
\subsection{Implementation}\label{implementation}}

The core functionality of this package is made available via the eponymous
\texttt{methyvim} function, which implements a statistical algorithm designed to
compute targeted estimates of VIMs, defined in such a way that the VIMs
represent parameters of scientific interest in computational biology
experiments; moreover, these VIMs are defined such that they may be estimated in
a manner that is very nearly assumption-free, that is, within a \emph{fully
nonparametric statistical model}. \textbf{The statistical algorithm consists in the
several major steps elaborated upon above and summarized below}:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  \emph{Pre-screening} of genomic sites is used to isolate a subset of sites for
  which there is cursory evidence of differential methylation. Currently, the
  available screening approach adapts core routines from the
  \href{http://bioconductor.org/packages/limma}{\texttt{limma}} R package. Following the
  style of the function for performing screening via \texttt{limma}, users may write
  their own screening functions and are invited to contribute such functions to
  the core software package by opening pull requests at the GitHub repository:
  \url{https://github.com/nhejazi/methyvim}.
\item
  Nonparametric estimates of VIMs, for the specified target parameter, are
  computed at each of the CpG sites passing the screening step. The VIMs are
  defined in such a way that the estimated effects is of an binary treatment on
  the methylation status of a target CpG site, controlling for the observed
  methylation status of the neighbors of that site. Currently, routines are
  adapted from the \href{https://CRAN.R-project.org/package=tmle}{\texttt{tmle}} R package.
\item
  Since pre-screening is performed prior to estimating VIMs, we apply the
  modified marginal Benjamini and Hochberg step-up False Discovery Rate
  controlling procedure for multi-stage analyses (FDR-MSA), which is
  well-suited for avoiding false positive discoveries when testing is only
  performed on a subset of potential targets.
\end{enumerate}

\hypertarget{parameters-of-interest}{%
\paragraph{Parameters of Interest}\label{parameters-of-interest}}

For the CpG sites that pass the pre-screening step, a user-specified target
parameter of interest is estimated independently at each site. \emph{In all cases,
an estimator of the parameter of interest is constructed via targeted minimum
loss-based estimation}.

For discrete-valued treatments or exposures:

\begin{itemize}
\item
  The average treatment effect (ATE): The effect of a binary exposure or
  treatment on the observed methylation at a target CpG site is estimated,
  controlling for the observed methylation at all other CpG sites in the same
  neighborhood as the target site, based on an additive form. Often denoted
  \(\psi_0 = \psi_0(1) - \psi_0(0)\), the parameter estimate represents the
  additive difference in methylation that would have been observed at the
  target site had all observations received the treatment versus the
  counterfactual under which none received the treatment.
\item
  The relative risk (RR): The effect of a binary exposure or treatment on the
  observed methylation at a target CpG site is estimated, controlling for the
  observed methylation at all other CpG sites in the same neighborhood as the
  target site, based on a geometric form. Often denoted,
  \(\psi_0 = \frac{\psi_0(1)}{\psi_0(0)}\), the parameter estimate represents the
  multiplicative difference in methylation that would have
  been observed at the target site had all observations received the treatment
  versus the counterfactual under which none received the treatment.
\end{itemize}

Estimating the VIM corresponding to the parameters above, for discrete-valued
treatments or exposures, requires two separate regression steps: one for the
treatment mechanism (propensity score) and one for the outcome regression.
Technical details on the nature of these regressions are discussed in
\citet{hernan2018causal}, and details for estimating these regressions in the
framework of targeted minimum loss-based estimation are discussed in
\citet{vdl2011targeted}.

\hypertarget{future-work}{%
\paragraph{Future Work}\label{future-work}}

Support for continuous-valued treatments or exposures is \emph{planned and ongoing}
but not yet available in stable releases of the \texttt{methyvim} software package.
Future releases will allow users to assess continuous-valued treatments by
relying on parameters estimable through implementations available in the
following software packages:

\begin{itemize}
\item
  A nonparametric variable importance measure (NPVI) \citep{chambaz2012estimation}
  (R package \texttt{tmle.npvi}). The effect of a continuous-valued exposure or
  treatment (the observed methylation at a target CpG site) on an outcome of
  interest is estimated, controlling for the observed methylation value at all
  other CpG sites in the same neighborhood as the target site. This uses a
  parameter that compares values of the treatment against a user-specified
  reference value taken to be the null value. In particular, the implementation
  to be provided is designed to assess the effect of differential methylation at
  the target CpG site on an outcome of interest (e.g., survival), providing a
  nonparametric evaluation of the impact of methylation at the target site.
\item
  The causal effect of shifting the value of an observed intervention, defined
  as the counterfactual outcome under a posited shift of a continuous-value
  treatment of interest using stochastic intervention policies
  \citep[\citet{diaz2018stochastic}, \citet{hejazi2018txshift}]{munoz2012population} (R package
  \texttt{txshift}). The value of an outcome of interest under an unobserved value of
  the (continuous-valued) treatment, specified through a user-provided
  \emph{additive shift} of the observed treatment, may be data-adaptively estimated.
  This allows for the effect of changes/shifts in the observed methylation at a
  CpG site on an outcome of interest to be evaluated, providing a nonparametric
  evalution of the relative importance of a particular target CpG site with
  respect to another variable measured in the same study.
\end{itemize}

\hypertarget{class-methytmle}{%
\paragraph{\texorpdfstring{Class \texttt{methytmle}}{Class methytmle}}\label{class-methytmle}}

We have adopted a class \texttt{methytmle} to help organize the functionality within
this package. The \texttt{methytmle} class builds upon the \texttt{GenomicRatioSet} class
provided by the \texttt{minfi} package so all of the slots of \texttt{GenomicRatioSet} are
contained in a \texttt{methytmle} object. The new class introduced in the \texttt{methyvim}
package includes several new slots:

\begin{itemize}
\item
  \texttt{call} - the form of the original call to the \texttt{methyvim} function.
\item
  \texttt{screen\_ind} - indices identifying CpG sites that pass the screening process.
\item
  \texttt{clusters} - non-unique IDs corresponding to the manner in wich sites are
  treated as neighbors. These are assigned by genomic distance (bp) and respect
  chromosome boundaries (produced via a call to \texttt{bumphunter::clusterMaker}).
\item
  \texttt{var\_int} - the treatment/exposure status for each subject. Currently, these
  must be binary, due to the definition of the supported targeted parameters.
\item
  \texttt{param} - the name of the target parameter from which the estimated VIMs are
  defined.
\item
  \texttt{vim} - a table of statistical results obtained from estimating VIMs for
  each of the CpG sites that pass the screening procedure.
\item
  \texttt{ic} - the measured array values for each of the CpG sites passing the
  screening, transformed into influence curve space based on the chosen target
  parameter.
\end{itemize}

The \texttt{show} method of the \texttt{methytmle} class summarizes a selection of the above
information for the user while masking some of the wealth of information given
when calling the same method for \texttt{GenomicRatioSet}. All information contained in
\texttt{GenomicRatioSet} objects is preserved in \texttt{methytmle} objects, so as to easen
interoperability with other differential methylation software for experienced
users. We refer the reader to the package vignette, ``\texttt{methyvim}: Targeted
Data-Adaptive Estimation and Inference for Differential Methylation Analysis,''
included in any distribution of the software package, for further details.

\hypertarget{operation}{%
\subsection{Operation}\label{operation}}

A standard computer with the latest version of R and Bioconductor 3.6 installed
will handle applications of the \texttt{methyvim} package.

\hypertarget{use-cases}{%
\section{Use Cases}\label{use-cases}}

To examine the practical applications and the full set of utilities of the
\texttt{methyvim} package, we will use a publicly available example data set produced
by the Illumina 450K array, from the \texttt{minfiData} R package.

\hypertarget{preliminaries-setting-up-the-data}{%
\paragraph{Preliminaries: Setting up the Data}\label{preliminaries-setting-up-the-data}}

We begin by loading the package and the data set. After loading the data, which
comes in the form of a raw \texttt{MethylSet} object, we perform some further
processing by mapping to the genome (with \texttt{mapToGenome}) and converting the
values from the methylated and unmethylated channels to Beta-values
(via \texttt{ratioConvert}). These two steps together produce an object of class
\texttt{GenomicRatioSet}, provided by the \texttt{minfi} package.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}
  \CommentTok{# numerous messages displayed at time of loading}
  \KeywordTok{library}\NormalTok{(minfiData)}
\NormalTok{)}
\KeywordTok{data}\NormalTok{(MsetEx)}
\NormalTok{mset <-}\StringTok{ }\KeywordTok{mapToGenome}\NormalTok{(MsetEx)}
\NormalTok{grs <-}\StringTok{ }\KeywordTok{ratioConvert}\NormalTok{(mset)}
\NormalTok{grs}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: GenomicRatioSet 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
\end{verbatim}

We can create an object of class \texttt{methytmle} from any \texttt{GenomicRatioSet} object
simply invoking the S4 class constructor \texttt{.methytmle}:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(methyvim)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## methyvim v1.3.1: Targeted Variable Importance for Differential Methylation Analysis
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{grs_mtmle <-}\StringTok{ }\KeywordTok{.methytmle}\NormalTok{(grs)}
\NormalTok{grs_mtmle}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: methytmle 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
## Target Parameter: 
## Results: 
## Object of class "data.frame"
## data frame with 0 columns and 0 rows
\end{verbatim}

Additionally, a \texttt{GenomicRatioSet} can be created from a matrix with the
function \texttt{makeGenomicRatioSetFromMatrix} provided by the \texttt{minfi} package.

\hypertarget{differential-methylation-analysis}{%
\paragraph{Differential Methylation Analysis}\label{differential-methylation-analysis}}

For this example analysis, we'll treat the condition of the patients as the
exposure/treatment variable of interest. The \texttt{methyvim} function requires that
this variable either be \texttt{numeric} or easily coercible to \texttt{numeric}. To
facilitate this, we'll simply convert the covariate (currently a \texttt{character}):

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{var_int <-}\StringTok{ }\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(}\KeywordTok{as.factor}\NormalTok{(}\KeywordTok{colData}\NormalTok{(grs)}\OperatorTok{$}\NormalTok{status)) }\OperatorTok{-}\StringTok{ }\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbf{n.b.}, the re-coding process results in ``normal'' patients being assigned a
value of 1 and cancer patients a 0.

Now, we are ready to analyze the effects of cancer status on DNA methylation
using this data set. We proceed as follows with a targeted minimum loss-based
estimate of the Average Treatment Effect.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{methyvim_cancer_ate <-}\StringTok{ }\KeywordTok{methyvim}\NormalTok{(}\DataTypeTok{data_grs =}\NormalTok{ grs, }\DataTypeTok{var_int =}\NormalTok{ var_int,}
                                \DataTypeTok{vim =} \StringTok{"ate"}\NormalTok{, }\DataTypeTok{type =} \StringTok{"Beta"}\NormalTok{, }\DataTypeTok{filter =} \StringTok{"limma"}\NormalTok{,}
                                \DataTypeTok{filter_cutoff =} \FloatTok{0.20}\NormalTok{, }\DataTypeTok{obs_per_covar =} \DecValTok{2}\NormalTok{,}
                                \DataTypeTok{parallel =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{sites_comp =} \DecValTok{250}\NormalTok{,}
                                \DataTypeTok{tmle_type =} \StringTok{"glm"}
\NormalTok{                               )}
\end{Highlighting}
\end{Shaded}

Note that we set the \texttt{obs\_per\_covar} argument to a relatively low value (just 2,
even though the recommended value, and default, is 20) for the purposes of this
example as the sample size is only 10. We do this only to exemplify the
estimation procedure and it is important to point out that such low values for
\texttt{obs\_per\_covar} will compromise the quality of inference obtained because this
setting directly affects the definition of the target parameter.

Further, note that here we apply the \texttt{glm} flavor of the \texttt{tmle\_type} argument,
which produces faster results by fitting models for the propensity score and
outcome regressions using a limited number of parametric models. By contrast,
the \texttt{sl} (for ``Super Learning'') flavor fits these two regressions using highly
nonparametric and data-adaptive procedures (i.e., via machine learning).
Obtaining the estimates via GLMs results in each of the regression steps
being less robust than if nonparametric regressions were used.

We can view a table of results by examining the \texttt{vim} slot of the produced
object, most easily displayed by simply printing the resultant object:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{methyvim_cancer_ate}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: methytmle 
## dim: 485512 6 
## metadata(0):
## assays(2): Beta CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(6): 5723646052_R02C02 5723646052_R04C01 ...
##   5723646053_R05C02 5723646053_R06C02
## colData names(13): Sample_Name Sample_Well ... Basename filenames
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.21.2
##   Manifest version: 0.4.0
## Target Parameter: Average Treatment Effect
## Results: 
##                   lwr_ci       est_ate        upr_ci      var_ate
## cg14008030 -1.195660e-01 -0.0314159619  0.0567340489 2.022705e-03
## cg20253340 -8.850637e-02 -0.0588661418 -0.0292259165 2.286919e-04
## cg21870274 -9.499057e-02 -0.0291189817  0.0367526091 1.129495e-03
## cg17308840 -4.626018e-02 -0.0071524518  0.0319552773 3.981191e-04
## cg00645010 -2.677328e-02 -0.0134655543 -0.0001578256 4.609945e-05
## cg27534567  6.745648e-02  0.1157118536  0.1639672225 6.061486e-04
## cg08258224  1.365045e-01  0.3050884951  0.4736724770 7.398105e-03
## cg20275697 -3.021026e-01 -0.1231299608  0.0558427144 8.337989e-03
## cg24373735 -4.283533e-02  0.0076666975  0.0581687266 6.639043e-04
## cg12445832 -6.082411e-02  0.0150395574  0.0909032203 1.498151e-03
## cg01097950 -4.392159e-02  0.0658323796  0.1755863507 3.135656e-03
## cg01782097 -1.082072e-02  0.0010232901  0.0128672954 3.651615e-05
##                    pval n_neighbors n_neighbors_control max_cor_neighbors
## cg14008030 4.848468e-01           0                   0                NA
## cg20253340 9.917426e-05           0                   0                NA
## cg21870274 3.862537e-01           2                   1        0.94435796
## cg17308840 7.199943e-01           2                   1        0.94435796
## cg00645010 4.734007e-02           2                   2        0.52368097
## cg27534567 2.602936e-06           1                   0        0.93629683
## cg08258224 3.895914e-04           1                   0        0.93629683
## cg20275697 1.775155e-01           0                   0                NA
## cg24373735 7.660489e-01           0                   0                NA
## cg12445832 6.976022e-01           0                   0                NA
## cg01097950 2.397377e-01           0                   0                NA
## cg01782097 8.655302e-01           1                   1       -0.39410834
##  [ reached getOption("max.print") -- omitted 238 rows ]
\end{verbatim}

Finally, we may compute FDR-corrected p-values, by applying a modified procedure
for controlling the False Discovery Rate for multi-stage analyses (FDR-MSA)
\citep{tuglus2009modified}. We do this by simply applying the \texttt{fdr\_msa} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fdr_p <-}\StringTok{ }\KeywordTok{fdr_msa}\NormalTok{(}\DataTypeTok{pvals =} \KeywordTok{vim}\NormalTok{(methyvim_cancer_ate)}\OperatorTok{$}\NormalTok{pval,}
                 \DataTypeTok{total_obs =} \KeywordTok{nrow}\NormalTok{(methyvim_cancer_ate))}
\end{Highlighting}
\end{Shaded}

Having explored the results of our analysis numerically, we now proceed to use
the visualization tools provided with the \texttt{methyvim} R package to further
enhance our understanding of the results.

\hypertarget{visualization-of-results}{%
\paragraph{Visualization of Results}\label{visualization-of-results}}

While making allowance for users to explore the full set of results produced by
the estimation procedure (by way of exposing these directly to the user), the
\texttt{methyvim} package also provides \emph{three} (3) visualization utilities that
produce plots commonly used in examining the results of differential methylation
analyses.

A simple call to \texttt{plot} produces side-by-side histograms of the raw p-values
computed as part of the estimation process and the corrected p-values obtained
from using the FDR-MSA procedure.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(methyvim_cancer_ate, }\DataTypeTok{type =} \StringTok{"raw_pvals"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-pvals-raw-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(methyvim_cancer_ate, }\DataTypeTok{type =} \StringTok{"fdr_pvals"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-pvals-fdr-1} \end{center}

\textbf{Remark:} The plots displayed above may also be generated as side-by-side
histograms in a single plot object. This is the default for the \texttt{plot} method
and may easily be invoked by specifying no additional arguments to the \texttt{plot}
function, unlike in the above.

While histograms of the p-values may be generally useful in inspecting the
results of the estimation procedure, a more common plot used in examining the
results of differential methylation procedures is the volcano plot, which plots
the parameter estimate along the x-axis and \(-\text{log}_{10}(\text{p-value})\)
along the y-axis. We implement such a plot in the \texttt{methyvolc} function:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{methyvolc}\NormalTok{(methyvim_cancer_ate)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-volcano-1} \end{center}

The purpose of such a plot is to ensure that very low (possibly statistically
significant) p-values do not arise from cases of low variance. This appears to
be the case in the plot above (notice that most parameter estimates are near
zero, even in cases where the raw p-values are quite low).

Yet another popular plot for visualizing effects in such settings is the
heatmap, which plots estimates of the raw methylation effects (as measured by
the assay) across subjects using a heat gradient. We implement this in the
\texttt{methyheat} function:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{methyheat}\NormalTok{(methyvim_cancer_ate, }\DataTypeTok{smooth.heat =} \OtherTok{TRUE}\NormalTok{, }\DataTypeTok{left.label =} \StringTok{"none"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.1 0.3 0.6 0.8 1.0
\end{verbatim}

\begin{center}\includegraphics{paper_BiocF1000_files/figure-latex/methyvim-heatmap-1} \end{center}

\textbf{Remark}: Invoking \texttt{methyheat} in this manner produces a plot of the top sites
(\(25\), by default) based on the raw p-value, using the raw methylation measures
in the plot. This uses the exceptional \texttt{superheat} R package
\citep{barter2017superheat}, to which we can easily pass additional parameters. In
particulat, we hide the CpG site labels that would appear by default on the left
of the heatmap (by setting \texttt{left.label\ =\ "none"}) to emphasize that this is only
an example and \emph{not} a scientific discovery.

\hypertarget{summary}{%
\section{Summary}\label{summary}}

Here we introduce the R package \texttt{methyvim}, an implementation of a general
algorithm for differential methylation analysis that allows for recent advances
in causal inference and machine learning to be leveraged in computational
biology settings. The estimation procedure produces straightforward statistical
inference and takes great care to ensure computationally efficiency of the
technique for obtaining targeted estimates of nonparametric variable importance
measures. The software package includes techniques for pre-screening a set of
CpG sites, controlling for the False Discovery Rate as if all sites were tested,
and for visualzing the results of the analyses in a variety of ways. The anatomy
of the software package is dissected and the design described in detail. The
\texttt{methyvim} R package is available via the Bioconductor project.

\hypertarget{software-availability}{%
\section{Software availability}\label{software-availability}}

Latest source code (development version): \url{https://github.com/nhejazi/methyvim}

Bioconductor (stable release): \url{https://bioconductor.org/packages/methyvim}

Archived source code as at time of publication:
\url{https://github.com/nhejazi/methyvim/releases/tag/f1000}

Documentation (development version): \url{https://code.nimahejazi.org/methyvim}

Software license: The MIT License, copyright Nima S. Hejazi

\hypertarget{author-contributions}{%
\section{Author contributions}\label{author-contributions}}

NH designed and implemented the software package, applied the tool to the use
cases presented, and co-drafted the present manuscript. RP helped in designing
the software and co-drafted the present manuscript. AH and ML served as advisors
for the development of this software and the general statistical algorithm it
implements.

\hypertarget{competing-interests}{%
\section{Competing interests}\label{competing-interests}}

No competing interests were disclosed at the time of publication.

\hypertarget{grant-information}{%
\section{Grant information}\label{grant-information}}

NH was supported in part by the National Library of Medicine of the National
Institutes of Health under Award Number T32-LM012417, by P42-ES004705, and
by R01-ES021369. RP was supported by P42-ES004705. The content of this work is
solely the responsibility of the authors and does not necessarily represent the
official views of the various funding sources and agencies.

\renewcommand\refname{References}
{\small\bibliography{references.bib}}

\end{document}
